README
SVG/TT/Graph
============

This set of modules can be used to generate high quality
SVG graphs, bar charts and pie graphs.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:
  DateTime >= 0.1
  HTTP::Date >= 0.1
  Template >= 2.08

CONVERTING TO PNG

Currently if you want to convert the SVG to PNG
I'd suggest using Batik from http://xml.apache.org/,
it is Java, I know there is work going on for the
perl converters, they're just not there yet.

DEVELOPMENT

The repository for the development of SVG::TT::Graph is:
   git://github.com/rafl/moosex-declare.git

COPYRIGHT AND LICENCE

This module is free software; you can redistribute it or 
modify it under the same terms as Perl itself.

Copyright (C) 2003 Leo Lapworth



Changes
Revision history for Perl extension SVG::TT::Graph.

1.00 Fri 10 May 2019 07:19:53
    - add support for `inline` (jmccanta)
    - doc and timezone fix (Eric Jensen)

0.26 Sun  4 Oct 2015 20:18:50
    - Add HeatMap! (stuartskelton)

0.25 Tue 14 Apr 2015 21:26:11
    - Make TimeSeries accept max_scale_value. (FGasper)
    - Teach TimeSeries to accept a time zone for date display. (FGasper)

0.24 Wed 9 Jul 2014 19:23:11
    - Added config.show_path_title to control whether to include new
      title attribute in path tag, default to '0' (mikejbwhoo)
    - Added TITLE attribute to PATH tag in SHOW DATA section (mikejbwhoo)
    - Added config.show_title_fields to allow for fields array element to
      be shown instead of title attribute used in add_data, looks nice
      when creating non-overlapping bars, default to '0' (mikejbwhoo)
    - Above changes made only to Pie, Bar, BarHorizontal, BarLine
      (mikejbwhoo)

0.23 Mon 4 Feb 2013 21:14:40
    - fix division by zero errors in all graph types (Carsten Wolff)
    - Doc clarification (elnjensen)

0.22 Sun 14 Oct 2012 09:43:58
    - Link to github for issue tracking
    - Allow reversed y-axis (elnjensen)
    - Add missing class for data point labels (snabb)

0.21
  - New type of graph available: XY (written by Florent Angly, but largely based
    on the TimeSeries module by David Meibusch)

0.20
  - Fixed wrong formatting of Y labels in SVG::TT:Graph::BarHorizontal (patch by
    Erin Schönhals)
  - Fixed wrong formatting of Y labels in SVG::TT::Graph::Bar

0.19
  - Fixed warning messages with missing data title (#69691, reported by
    dean@fragfest.com.au, patch by Florent Angly)

0.18 - Thu  5 Jan 2012
  - Added format_x_axis, format_y_axis callback options (Eric Johnson)

0.17 - Wed Jul 20 2011
  - By default, do not compress() or tidy() the SVG
  - Line and Timeseries graphs now have points styled as '.dataPointNN',
    consistently with BarLine graphs (patch from Henrik Pauli, #67834)
  - Added a max_scale_value option to set the maximum value for the Y axis of
    Bar and Line graphs (patch by Bryce and Henrik Pauli, #20475)
  - Repackaged module without extra files (#56718)

0.16 - Sun Apr 18 2010
  - Javascript code in the piechart and timeseries put into a CDATA tag

0.15 - Sat Apr 17 2010
  - Method get_template() moved from SVG::TT::Graph::* to SVG::TT::Graph to avoid redundancy
  - Fixed bug in get_template(). The DATA filehandle is now read from the beginning
    at each get_template call
  - Makefile.PL does not use autoinstall aymore (not a good practice for CPAN modules)
  - Rollover improvements in piechart

0.14 - Mon Apr 05 2010
  - Fixed some unit tests and the URL of the repository

0.13 - Sat Apr 03 2010
  - Revised TT templates warnings due to undef values (bug # 44115)
  - Fixed the "used only once" warnings (bug # 43854, thank you Ia Hun)
  - Introduced a new method to use random colors instead of the 12 predefined
    colors. Along with the dynamic allocation of the number of style elements,
    this allows an unlimited number of dataseries
  - Added Javascript code to display field name when the mouse rolls over a
    piechart wedge
  - Implemented an optional new method to tidy the formatting of the SVG files
    (using XML::Tidy)
  - Replaced carriage returns by line feeds because it is more portable
    (http://www.webmasterworld.com/forum13/4067.htm)
  - Did documentation, unit testing and packaging revisions
  - Misc little code touch ups

0.12 - Mon Feb 16 12:00:00 EST 2009
  -  Missed updating the versions for some of the sub libraries.

0.11 - Sun Feb 15 22:54:00 EST 2009
  - Applied patches:
    http://rt.cpan.org/Ticket/Display.html?id=38441 - Ampersands in Fields cause XML error
    http://rt.cpan.org/Ticket/Display.html?id=43326 - Multiple data sets in Bar.pm will now show stacked graphs
    http://rt.cpan.org/Ticket/Display.html?id=43325 - Allow toggling the display of data on TimeSeries

0.10 - Mon Apr 28 21:07:30 BST 2008
  - Applied patches:
    http://rt.cpan.org/Ticket/Display.html?id=34491
    http://rt.cpan.org/Ticket/Display.html?id=22896
    http://rt.cpan.org/Ticket/Display.html?id=20828
    http://rt.cpan.org/Ticket/Display.html?id=20475

0.09
  - Applied path patch for Firefox from Pat Ekman for Pie
  - patched other modules so Firefox works (doesn't like comma's in a path)

0.08
  - Fix stylesheet for Pie

0.07
  Help from Yanick with the following:

  - Bug fixes:
    #7342 (bad default stylesheet class)
    #8680 (Typo in defaut Stylesheet for SVG::TT::Graph::TimeSeries)
    #7541 (BarHorizontal)
    #7549 (Character width setting in BarHorizontal)

0.06
  - Put in comment so people know where to copy the
    style sheet from (e.g. not including the <style> or CDATA)

0.05 (dev)
  David Meibusch changes:
  - Calls calculations() prior to burn if available
  - 'calc' now passed to template if available
  - _range_calc has moved from TimeSeries.pm to Graph.pm

  TimeSeries:
  - rollover_values
  - data_value_format
  - max_time_span
  - y_label_format
  - rotate_x_labels

  Other changes:
  - Renamed README to INSTALL
  - rotate_x_labels added
  - y_title_text_direction added
  - scale_integers added
  - Minor doc patch "\r\n" to "\r\n\r\n"
  - Stylesheet presets increased (from 10) to support 12 datasets
  - Bug fix regarding show_data_values with Bar
  - Bug fix with scales of values under 5
  - Added BarLine as a graph module
  - Added TimeSeries as a graph module - by David
  - Added Prereq of 'DateTime' - lots of modules but
    don't think it's bad to encourage people to use it
  - Altered style sheet from fill-opacity: 0.2; to fill-opacity: 0.5;

0.04 Wed Sept 24 2003
  - Patch from David Meibusch for Pie if slice > 50%
  - Patch from David Meibusch for Line to add 'stacked' option

0.03  Wed Aug 28
  - Doc fix thanks to Mike Castle
  - Renaming (sorry, thought it best to do ASAP)
  - y_marker  to  scale_divisions
  - xfields   to  fields
  - y_start   to  min_scale_value

lib/SVG/TT/Graph.pm
package SVG::TT::Graph;

use strict;
use Carp;
use Template;

require 5.6.1;

our $VERSION = '0.26';
our $AUTOLOAD;
our $TEMPLATE_FH;

=head1 NAME

SVG::TT::Graph - Base module for generating SVG graphics

=head1 SYNOPSIS

  package SVG::TT::Graph::GRAPH_TYPE;
  use SVG::TT::Graph;
  use base qw(SVG::TT::Graph);
  our $VERSION = $SVG::TT::Graph::VERSION;
  our $TEMPLATE_FH = \*DATA;

  sub _set_defaults {
    my $self = shift;

    my %default = (
        'keys'  => 'value',
    );
    while( my ($key,$value) = each %default ) {
      $self->{config}->{$key} = $value;
    }
  }


  # optional - called when object is created
  sub _init {
    my $self = shift;
  # any testing you want to do.

  }

  ...

  1;
  __DATA__
  <!-- SVG Template goes here  -->


  In your script:

  use SVG::TT::Graph::GRAPH_TYPE;

  my $width = '500',
  my $heigh = '300',
  my @fields = qw(field_1 field_2 field_3);

  my $graph = SVG::TT::Graph::GRAPH_TYPE->new({
    # Required for some graph types
    'fields'           => \@fields,
    # .. other config options
    'height' => '500',
  });

  my @data = qw(23 56 32);
  $graph->add_data({
    'data' => \@data,
    'title' => 'Sales 2002',
  });

  # find a config options value
  my $config_value = $graph->config_option();
  # set a config option value
  $graph->config_option($config_value);

  # All graphs support SVGZ (compressed SVG) if
  # Compress::Zlib is available. Use either the
  # 'compress' => 1 config option, or:
  $graph->compress(1);

  # All graph SVGs can be tidied if XML::Tidy
  # is installed. Use either the 'tidy' => 1
  # config option, or:
  $graph->tidy(1);

  print "Content-type: image/svg+xml\n\n";
  print $graph->burn();

=head1 DESCRIPTION

This package is a base module for creating graphs in Scalable Vector Format
(SVG). Do not use this module directly. Instead, use one of the following
modules to create the plot of your choice:

=over

=item L<SVG::TT::Graph::Line>

=item L<SVG::TT::Graph::Bar>

=item L<SVG::TT::Graph::BarHorizontal>

=item L<SVG::TT::Graph::BarLine>

=item L<SVG::TT::Graph::Pie>

=item L<SVG::TT::Graph::TimeSeries>

=item L<SVG::TT::Graph::XY>

=back

If XML::Tidy is installed, the SVG files generated can be tidied. If
Compress::Zlib is available, the SVG files can also be compressed to SVGZ.

=cut

sub new {
  my ($proto,$conf) = @_;
  my $class = ref($proto) || $proto;
  my $self = {};

  bless($self, $class);

  if($self->can('_set_defaults')) {
    # Populate with local defaults
    $self->_set_defaults();
  } else {
    croak "$class should have a _set_defaults method";
  }

  # overwrite defaults with user options
  while( my ($key,$value) = each %{$conf} ) {
    $self->{config}->{$key} = $value;
  }

  # Allow the inheriting modules to do checks
  if($self->can('_init')) {
    $self->_init();
  }

  return $self;
}

=head1 METHODS

=head2 add_data()

  my @data_sales_02 = qw(12 45 21);

  $graph->add_data({
    'data' => \@data_sales_02,
    'title' => 'Sales 2002',
  });

This method allows you do add data to the graph object.
It can be called several times to add more data sets in.

=cut

sub add_data {
  my ($self, $conf) = @_;
  # create an array
  unless(defined $self->{'data'}) {
    my @data;
    $self->{'data'} = \@data;
  }

  croak 'no fields array ref'
  unless defined $self->{'config'}->{'fields'}
    && ref($self->{'config'}->{'fields'}) eq 'ARRAY';

  if(defined $conf->{'data'} && ref($conf->{'data'}) eq 'ARRAY') {
    my %new_data;
    @new_data{ map { s/&/&amp;/; $_ } @{$self->{'config'}->{'fields'}}} = @{$conf->{'data'}};
    my %store = (
      'data' => \%new_data,
    );
    $store{'title'} = $conf->{'title'} if defined $conf->{'title'};
    push (@{$self->{'data'}},\%store);
    return 1;
  }
  return undef;
}

=head2 clear_data()

  my $graph->clear_data();

This method removes all data from the object so that you can
reuse it to create a new graph but with the same config options.

=cut

sub clear_data {
  my $self = shift;
  my @data;
  $self->{'data'} = \@data;
}


=head2 get_template()

  print $graph->get_template();

This method returns the TT template used for making the graph.

=cut

sub get_template {
  my $self = shift;

  # Template filehandle
  my $template_fh_sr = $self->_get_template_fh_sr();
  croak ref($self) . ' must have a template' if not $template_fh_sr;

  my $template_fh = $$template_fh_sr;

  # Read in TT template
  my $start = tell $template_fh;
  my $template = '';
  while(<$template_fh>) {
    chomp;
    $template .= $_ . "\n";
  }

  # This method may be used again, so return to start of filehandle
  seek $template_fh, $start, 0;

  return $template;
}

sub _get_template_fh_sr {
    my ($self) = @_;

    my $ns_ref = \%main::;
    for my $node ( split m<::>, ref $self ) {
        $ns_ref = $ns_ref->{"${node}::"};
    }

    return *{$ns_ref->{'TEMPLATE_FH'}}{'SCALAR'};
}

=head2 burn()

  print $graph->burn();

This method processes the template with the data and
config which has been set and returns the resulting SVG.

This method will croak unless at least one data set has
been added to the graph object.

=cut

sub burn {
  my $self = shift;

  # Check we have at least one data value
  croak "No data available"
    unless scalar(@{$self->{'data'}}) > 0;

  # perform any calculations prior to burn
  $self->calculations() if $self->can('calculations');

  my $template = $self->get_template();
  my %vals = (
    'data'             => $self->{'data'},     # the data
    'config'           => $self->{'config'},   # the configuration
    'calc'             => $self->{'calc'},     # the calculated values
    'sin'              => \&_sin_it,
    'cos'              => \&_cos_it,
    'predefined_color' => \&_predefined_color,
    'random_color'     => \&_random_color,
  );

  # euu - hack!! - maybe should just be a method
  $self->{sin} = \&_sin_it;
  $self->{cos} = \&_cos_it;

  # set up TT object
  my %config = (
    POST_CHOMP   => 1,
    INCLUDE_PATH => '/',
    #STRICT      => 1, # we should probably enable this for strict checking
    #DEBUG       => 'undef', # TT warnings on, useful for debugging, finding undef values
  );
  my $tt = Template->new( \%config );
  my $file;

  my $template_response = $tt->process( \$template, \%vals, \$file );
  if($tt->error()) {
    croak "Template error: " . $tt->error . "\n" if $tt->error;
  }

  # tidy SVG if required
  if ($self->tidy()) {
    if (eval "require XML::Tidy") {
      # remove the doctype tag temporarily because it seems to cause trouble
      $file =~ s/(<!doctype svg .+?>)//si;
      my $doctype = $1;
      # tidy
      my $tidy_obj = XML::Tidy->new( 'xml' => $file );
      $tidy_obj->tidy();
      $file = $tidy_obj->toString();
      # re-add the doctype
      if (defined $doctype) {
        $file =~ s/(<\?xml.+?\?>)/$1\n$doctype/si;
      }
      # even more tidy
      $file = $self->_tidy_more($file);
    } else {
      croak "Error tidying the SVG file: XML::Tidy does not seem to be installed properly";
    }
  }

  # compress SVG if required
  if ($self->compress()) {
    if (eval "require Compress::Zlib") {
      $file = Compress::Zlib::memGzip($file);
    } else {
      croak "Error compressing the SVG file: Compress::Zlib does not seem to be installed properly";
    }
  }

  return $file;
}



sub _sin_it {
  return sin(shift);
}


sub _cos_it {
  return cos(shift);
}


=head2 compress()

  $graph->compress(1);

If Compress::Zlib is installed, the content of the SVG file can be compressed.
This get/set method controls whether or not to compress. The default is 0 (off).

=cut

sub compress {
  my ($self, $val) = @_;
  # set the default compress value
  if (not defined $self->{config}->{compress}) {
    $self->{config}->{compress} = 0;
  }
  # set the user-defined compress value
  if (defined $val) {
    $self->{config}->{compress} = $val;
  }
  # get the compress value
  return $self->{config}->{compress};
}


=head2 tidy()

  $graph->tidy(1);

If XML::Tidy is installed, the content of the SVG file can be formatted in a
prettier way. This get/set method controls whether or not to tidy. The default
is 0 (off). The limitations of tidy are described at this URL:
L<http://search.cpan.org/~pip/XML-Tidy-1.12.B55J2qn/Tidy.pm#tidy%28%29>

=cut

sub tidy {
  my ($self, $val) = @_;
  # set the default tidy value
  if (not defined $self->{config}->{tidy}) {
    $self->{config}->{tidy} = 0;
  }
  # set the user-defined tidy value
  if (defined $val) {
    $self->{config}->{tidy} = $val;
  }
  # get the tidy value
  return $self->{config}->{tidy};
}


sub _tidy_more {
  # Remove extra spaces in the SVG <path> tag
  my ($self, $svg_string) = @_;
  while ($svg_string =~ s/(<path .*? )\s+(.*?"\s*?\/>)/$1$2/mgi) {};
  return $svg_string;
}


sub _random_color {
  # Generate the rgb code for a randomly selected color
  my $rgb = 'rgb('.int(rand(256)).','.int(rand(256)).','.int(rand(256)).')';
  return $rgb;
}


sub _predefined_color {
  # Get the hexadecimal code for one of 12 predefined colors
  my ($num) = shift;
  my @colors = ("#ff0000", "#0000ff", "#00ff00", "#ffcc00", "#00ccff",
    "#ff00ff", "#00ffff", "#ffff00", "#cc6666", "#663399", "#339900", "#9966FF");
  my $hex;
  if ($num-1 < scalar @colors) {
    $hex = $colors[$num-1];
  }
  return $hex;
}

# Calculate a scaling range and divisions to be aesthetically pleasing
# Parameters:
#   value range
# Returns
#   (revised range, division size, division precision)
sub _range_calc () {
  my ($self, $range) = @_;

  my ($max,$division);
  my $count = 0;
  my $value = $range;

  if ($value == 0) {
    # Can't do much really
    $division = 0.2;
    $max = 1;
    return ($max,$division,1);
  }

  if (($value < 1) and ($value > 0)) {
    while ($value < 1) {
      $value *= 10;
      $count++;
    }
    $division = 1;
    while ($count--) {
      $division /= 10;
    }
    $max = ceil($range / $division) * $division;
  }
  else {
    while ($value > 10) {
      $value /= 10;
      $count++;
    }
    $division = 1;
    while ($count--) {
      $division *= 10;
    }
    $max = ceil($range / $division) * $division;
  }

  if (int($max / $division) <= 2) {
    $division /= 5;
    $max = ceil($range / $division) * $division;
  }
  elsif (int($max / $division) <= 5) {
    $division /= 2;
    $max = ceil($range / $division) * $division;
  }

  if ($division >= 1) {
    $count = 0;
  }
  else {
    $count = length($division) - 2;
  }

  return ($max,$division,$count);
}


# Returns true if config value exists, is defined and not ''
sub _is_valid_config() {
  my ($self,$name) = @_;
  return ((exists $self->{config}->{$name}) && (defined $self->{config}->{$name}) && ($self->{config}->{$name} ne ''));
}


=head2 config methods

  my $value = $graph->method();
  $graph->method($value);

This object provides autoload methods for all config
options defined in the _set_default method within the
inheriting object.

See the SVG::TT::Graph::GRAPH_TYPE documentation for a list.

=cut

## AUTOLOAD FOR CONFIG editing

sub AUTOLOAD {
  my $name = $AUTOLOAD;
  $name =~ s/.*://;

  croak "No object supplied" unless $_[0];
  if(defined $_[0]->{'config'}->{$name}) {
    if(defined $_[1]) {
      # set the value
      $_[0]->{'config'}->{$name} = $_[1];
    }
    return $_[0]->{'config'}->{$name} if defined $_[0]->{'config'}->{$name};
    return undef;
  } else {
    croak "Method: $name can not be used with " . ref($_[0]);
  }
}


# As we have AUTOLOAD we need this
sub DESTROY {
}


1;
__END__

=head1 EXAMPLES

For examples look at the project home page http://leo.cuckoo.org/projects/SVG-TT-Graph/

=head1 EXPORT

None by default.

=head1 ACKNOWLEDGEMENTS

Thanks to Foxtons for letting us put this on CPAN, Todd Caine for heads up on
reparsing the template (but not using atm), David Meibusch for TimeSeries and a
load of other ideas, Stephen Morgan for creating the TT template and SVG, and
thanks for all the patches by Andrew Ruthven and others.

=head1 AUTHOR

Leo Lapworth <LLAP@cuckoo.org>

=head1 MAINTAINER

Florent Angly <florent.angly@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003, Leo Lapworth

This module is free software; you can redistribute it or
modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<SVG::TT::Graph::Line>,
L<SVG::TT::Graph::Bar>,
L<SVG::TT::Graph::BarHorizontal>,
L<SVG::TT::Graph::BarLine>,
L<SVG::TT::Graph::Pie>,
L<SVG::TT::Graph::TimeSeries>,
L<SVG::TT::Graph::XY>,
L<Compress::Zlib>,
L<XML::Tidy>

=cut
